### ISSUE TITLE
Too many open file error (Tomcat), if fluentd server is down

### ISSUE THREAD

**Issue by danishkhan121** [komamitsu/fluency#132]
Config

		FluencyBuilderForFluentd builder = new FluencyBuilderForFluentd();
		builder.setBufferChunkInitialSize(16 * 1024 * 1024);
		builder.setBufferChunkRetentionSize(64 * 1024 * 1024);
		builder.setMaxBufferSize(1024 * 1024 * 1024L);
		builder.setFlushIntervalMillis(10000);
		fluency = builder.build("xxxxxxxxx", 24224);

Trace:

2019-05-09 10:40:16,588  WARN [pool-2-thread-1] o.k.f.f.i.s.RetryableSender [RetryableSender.java:86] Sender failed to send data. sender=RetryableSender{baseSender=TCPSender{config=Config{host='xxxxxxxxxxxx', port=24224, connectionTimeoutMilli=5000, readTimeoutMilli=5000, waitBeforeCloseMilli=1000} Config{senderErrorHandler=null}} NetworkSender{config=Config{host='xxxxxxxxxxxx', port=24224, connectionTimeoutMilli=5000, readTimeoutMilli=5000, waitBeforeCloseMilli=1000} Config{senderErrorHandler=null}, failureDetector=null} org.komamitsu.fluency.fluentd.ingester.sender.TCPSender@2e130737, retryStrategy=ExponentialBackOffRetryStrategy{config=Config{baseIntervalMillis=400, maxIntervalMillis=30000} Config{maxRetryCount=7}} RetryStrategy{config=Config{baseIntervalMillis=400, maxIntervalMillis=30000} Config{maxRetryCount=7}}, isClosed=false} org.komamitsu.fluency.fluentd.ingester.sender.RetryableSender@25eac167, retry=3
java.net.SocketException: Too many open files
	at sun.nio.ch.Net.socket0(Native Method)
	at sun.nio.ch.Net.socket(Net.java:411)
	at sun.nio.ch.Net.socket(Net.java:404)

**Comment by komamitsu** [comment]
Hi @danishkhan121,

Could you give me the following information?

- OS (including version)
- JDK version
- When/How often does the application create Fluency instance(s)?
- How long did the Fluentd down?
- The output of `lsof -p $(the Java application's PID)` if you can reproduce it
  - Please mask sensitive information
- Error messages that were output before the above exception

**Comment by danishkhan121** [comment]
- OS (including version)

No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 14.04.4 LTS
Release:	14.04
Codename:	trusty

- JDK version

Spring boot 2.0 application running on Java 1.8 with internal tomcat

- When/How often does the application create Fluency instance(s)?

Single instance on startup (as logback appender) -attached

- How long did the Fluentd down?

The server was never up on given host 
[trace.txt](https://github.com/komamitsu/fluency/files/3165026/trace.txt)
[lsof.txt](https://github.com/komamitsu/fluency/files/3165027/lsof.txt)
[attachments.zip](https://github.com/komamitsu/fluency/files/3165034/attachments.zip)




- The output of lsof -p $(the Java application's PID) if you can reproduce it

Attached (Application was restarted however attaching current state (close to 158 open files)).This will keep increasing until limit is reached

- Error messages that were output before the above exception

Attached

**Comment by komamitsu** [comment]
Thanks to the error log and the result of `lsof` you attached, I got the cause.

According to the error messages, many `java.net.UnknownHostException` occurred. I'm not sure it was due to infrastructure side issue. But anyway I noticed Fluency could leak sockets when the exception was keeping to happen.

I usually test this kind of leak with `java.net.ConnectException: Connection refused`, but it doesn't reproduce this leak...

I'll fix this issue soon.
